File: hu/codes/java/chapter_stack_and_queue/linkedlist_deque.java
/**
 * File: linkedlist_deque.java
 * Created Time: 2023-01-20
 * Author: krahets (krahets@163.com)
 */

package chapter_stack_and_queue;

import java.util.*;

/* Kétirányú láncolt lista csomópont */
class ListNode {
    int val; // Csomópont értéke
    ListNode next; // Következő csomópont referenciája
    ListNode prev; // Előző csomópont referenciája

    ListNode(int val) {
        this.val = val;
        prev = next = null;
    }
}

/* Kétirányú láncolt lista alapú kétirányú sor implementáció */
class LinkedListDeque {
    private ListNode front, rear; // Fejcsomópont front, farokcsomópont rear
    private int queSize = 0; // Kétirányú sor hossza

    public LinkedListDeque() {
        front = rear = null;
    }

    /* Kétirányú sor hosszának lekérése */
    public int size() {
        return queSize;
    }

    /* Kétirányú sor üresség ellenőrzése */
    public boolean isEmpty() {
        return size() == 0;
    }

    /* Besorolási művelet */
    private void push(int num, boolean isFront) {
        ListNode node = new ListNode(num);
        // Ha a láncolt lista üres, front és rear is a csomópontra mutat
        if (isEmpty())
            front = rear = node;
        // Sor elejére való besorolási művelet
        else if (isFront) {
            // Csomópont hozzáadása a láncolt lista elejéhez
            front.prev = node;
            node.next = front;
            front = node; // Fejcsomópont frissítése
        // Sor végére való besorolási művelet
        } else {
            // Csomópont hozzáadása a láncolt lista végéhez
            rear.next = node;
            node.prev = rear;
            rear = node; // Farokcsomópont frissítése
        }
        queSize++; // Sor hosszának frissítése
    }

    /* Sor elejére való besorolás */
    public void pushFirst(int num) {
        push(num, true);
    }

    /* Sor végére való besorolás */
    public void pushLast(int num) {
        push(num, false);
    }

    /* Kisorolási művelet */
    private int pop(boolean isFront) {
        if (isEmpty())
            throw new IndexOutOfBoundsException();
        int val;
        // Fejcsomópont értékének ideiglenes tárolása
        if (isFront) {
            val = front.val; // Fejcsomópont törlése
            // Fejcsomópont törlése
            ListNode fNext = front.next;
            if (fNext != null) {
                fNext.prev = null;
                front.next = null;
            }
            front = fNext; // Fejcsomópont frissítése
        // Farokcsomópont értékének ideiglenes tárolása
        } else {
            val = rear.val; // Farokcsomópont törlése
            // Farokcsomópont frissítése
            ListNode rPrev = rear.prev;
            if (rPrev != null) {
                rPrev.next = null;
                rear.prev = null;
            }
            rear = rPrev; // Farokcsomópont frissítése
        }
        queSize--; // Sor hosszának frissítése
        return val;
    }

    /* Sor végéről való kisorolás */
    public int popFirst() {
        return pop(true);
    }

    /* Sor végén lévő elem elérése */
    public int popLast() {
        return pop(false);
    }

    /* Lista visszaadása nyomtatáshoz */
    public int peekFirst() {
        if (isEmpty())
            throw new IndexOutOfBoundsException();
        return front.val;
    }

    /* Főprogram */
    public int peekLast() {
        if (isEmpty())
            throw new IndexOutOfBoundsException();
        return rear.val;
    }

    /* Tömb visszaadása nyomtatáshoz */
    public int[] toArray() {
        ListNode node = front;
        int[] res = new int[size()];
        for (int i = 0; i < res.length; i++) {
            res[i] = node.val;
            node = node.next;
        }
        return res;
    }
}

public class linkedlist_deque {
    public static void main(String[] args) {
        /* Kétirányú sor hosszának lekérése */
        LinkedListDeque deque = new LinkedListDeque();
        deque.pushLast(3);
        deque.pushLast(2);
        deque.pushLast(5);
        System.out.println("Double-ended queue deque = " + Arrays.toString(deque.toArray()));

        /* Elem frissítése */
        int peekFirst = deque.peekFirst();
        System.out.println("Front element peekFirst = " + peekFirst);
        int peekLast = deque.peekLast();
        System.out.println("Rear element peekLast = " + peekLast);

        /* Elemek sorba rakása */
        deque.pushLast(4);
        System.out.println("After element 4 enqueues at rear, deque = " + Arrays.toString(deque.toArray()));
        deque.pushFirst(1);
        System.out.println("After element 1 enqueues at front, deque = " + Arrays.toString(deque.toArray()));

        /* Elem kivétele */
        int popLast = deque.popLast();
        System.out.println("Rear dequeue element = " + popLast + ", after rear dequeue, deque = " + Arrays.toString(deque.toArray()));
        int popFirst = deque.popFirst();
        System.out.println("Front dequeue element = " + popFirst + ", after front dequeue, deque = " + Arrays.toString(deque.toArray()));

        /* Kétirányú sor hosszának lekérése */
        int size = deque.size();
        System.out.println("Double-ended queue length size = " + size);

        /* Kétirányú sor üresség ellenőrzése */
        boolean isEmpty = deque.isEmpty();
        System.out.println("Double-ended queue is empty = " + isEmpty);
    }
}

File: hu/codes/python/chapter_stack_and_queue/linkedlist_deque.py
"""
File: linkedlist_deque.py
Created Time: 2023-03-01
Author: krahets (krahets@163.com)
"""


class ListNode:
    """Kétirányú láncolt lista csomópontja"""

    def __init__(self, val: int):
        """Konstruktor"""
        self.val: int = val
        self.next: ListNode | None = None  # Következő csomópontra mutató referencia
        self.prev: ListNode | None = None  # Előző csomópontra mutató referencia


class LinkedListDeque:
    """Kétirányú láncolt listán alapuló kétirányú sor implementáció"""

    def __init__(self):
        """Konstruktor"""
        self._front: ListNode | None = None  # Fejcsomópont (sor eleje)
        self._rear: ListNode | None = None  # Farokcsomópont (sor vége)
        self._size: int = 0  # A kétirányú sor hossza

    def size(self) -> int:
        """A kétirányú sor hosszának lekérése"""
        return self._size

    def is_empty(self) -> bool:
        """Ellenőrzi, hogy a kétirányú sor üres-e"""
        return self._size == 0

    def push(self, num: int, is_front: bool):
        """Sorba állás művelet"""
        node = ListNode(num)
        # Ha a láncolt lista üres, mind a front, mind a rear a csomópontra mutat
        if self.is_empty():
            self._front = self._rear = node
        # Sor elejéhez való sorba állás
        elif is_front:
            # Csomópont hozzáadása a láncolt lista elejéhez
            self._front.prev = node
            node.next = self._front
            self._front = node  # Fejcsomópont frissítése
        # Sor végéhez való sorba állás
        else:
            # Csomópont hozzáadása a láncolt lista végéhez
            self._rear.next = node
            node.prev = self._rear
            self._rear = node  # Farokcsomópont frissítése
        self._size += 1  # Sor hosszának frissítése

    def push_first(self, num: int):
        """Sor elejére való sorba állás"""
        self.push(num, True)

    def push_last(self, num: int):
        """Sor végére való sorba állás"""
        self.push(num, False)

    def pop(self, is_front: bool) -> int:
        """Sorból kivétel művelet"""
        if self.is_empty():
            raise IndexError("A kétirányú sor üres")
        # Sor elejéről való kivétel
        if is_front:
            val: int = self._front.val  # Fejcsomópont értékének ideiglenes tárolása
            # Fejcsomópont törlése
            fnext: ListNode | None = self._front.next
            if fnext is not None:
                fnext.prev = None
                self._front.next = None
            self._front = fnext  # Fejcsomópont frissítése
        # Sor végéről való kivétel
        else:
            val: int = self._rear.val  # Farokcsomópont értékének ideiglenes tárolása
            # Farokcsomópont törlése
            rprev: ListNode | None = self._rear.prev
            if rprev is not None:
                rprev.next = None
                self._rear.prev = None
            self._rear = rprev  # Farokcsomópont frissítése
        self._size -= 1  # Sor hosszának frissítése
        return val

    def pop_first(self) -> int:
        """Sor elejéről való kivétel"""
        return self.pop(True)

    def pop_last(self) -> int:
        """Sor végéről való kivétel"""
        return self.pop(False)

    def peek_first(self) -> int:
        """A sor elejének eleme"""
        if self.is_empty():
            raise IndexError("A kétirányú sor üres")
        return self._front.val

    def peek_last(self) -> int:
        """A sor végének eleme"""
        if self.is_empty():
            raise IndexError("A kétirányú sor üres")
        return self._rear.val

    def to_array(self) -> list[int]:
        """Tömb visszaadása nyomtatáshoz"""
        node = self._front
        res = [0] * self.size()
        for i in range(self.size()):
            res[i] = node.val
            node = node.next
        return res


"""Fő kód"""
if __name__ == "__main__":
    # Kétirányú sor inicializálása
    deque = LinkedListDeque()
    deque.push_last(3)
    deque.push_last(2)
    deque.push_last(5)
    print("kétirányú sor deque =", deque.to_array())

    # Elemek elérése
    peek_first: int = deque.peek_first()
    print("Sor eleje peek_first =", peek_first)
    peek_last: int = deque.peek_last()
    print("Sor vége peek_last =", peek_last)

    # Elemek sorba állítása
    deque.push_last(4)
    print("A 4-es elem sor végéhez adása után deque =", deque.to_array())
    deque.push_first(1)
    print("Az 1-es elem sor elejéhez adása után deque =", deque.to_array())

    # Elemek kivétele
    pop_last: int = deque.pop_last()
    print("Sor végéről kivett elem =", pop_last, ", sor végéről kivétel után deque =", deque.to_array())
    pop_first: int = deque.pop_first()
    print("Sor elejéről kivett elem =", pop_first, ", sor elejéről kivétel után deque =", deque.to_array())

    # A kétirányú sor hosszának lekérése
    size: int = deque.size()
    print("Kétirányú sor hossza size =", size)

    # Ellenőrzi, hogy a kétirányú sor üres-e
    is_empty: bool = deque.is_empty()
    print("A kétirányú sor üres-e =", is_empty)
